src/uiHelpers.h

```diff
@@ -8,71 +8,126 @@
 } ui_callback_t;
 
 
 typedef struct {
-	uint16_t initMagic;
-	char header[30];
-	char currentText[18];
-	char fullText[200];
-	size_t scrollIndex;
-	ui_callback_t callback;
-	#ifdef HEADLESS
-	bool headlessShouldRespond;
-	#endif
+    uint16_t initMagic;
+    char header[30];
+    char currentText[18];
+    char fullText[200];
+    size_t scrollIndex;
+    ui_callback_t callback;
+#ifdef HEADLESS
+    bool headlessShouldRespond;
+#endif
 } scrollingState_t;
 
 typedef struct {
-	uint16_t initMagic;
-	char header[30];
-	char text[30];
-	ui_callback_t callback;
-	#ifdef HEADLESS
-	bool headlessShouldRespond;
-	#endif
+    uint16_t initMagic;
+    char header[30];
+    char text[30];
+    ui_callback_t callback;
+#ifdef HEADLESS
+    bool headlessShouldRespond;
+#endif
 } confirmState_t;
 
 typedef union {
-	scrollingState_t scrolling;
-	confirmState_t confirm;
+    scrollingState_t scrolling;
+    confirmState_t confirm;
 } displayState_t;
 
 
 void ui_idle(void);
 
 void ui_displayScrollingText(
-        const char* headerStr,
-        const char* bodyStr,
-        ui_callback_fn_t* callback);
+    const char* headerStr,
+    const char* bodyStr,
+    ui_callback_fn_t* callback);
 
 void ui_displayConfirm(
-        const char* headerStr,
-        const char* bodyStr,
-        ui_callback_fn_t* confirm,
-        ui_callback_fn_t* reject
+    const char* headerStr,
+    const char* bodyStr,
+    ui_callback_fn_t* confirm,
+    ui_callback_fn_t* reject
 );
 
 void ui_checkUserConsent(
-        security_policy_t policy,
-        const char* headerStr,
-        const char* bodyStr,
-        ui_callback_fn_t* confirm,
-        ui_callback_fn_t* reject
+    security_policy_t policy,
+    const char* headerStr,
+    const char* bodyStr,
+    ui_callback_fn_t* confirm,
+    ui_callback_fn_t* reject
 );
 
 void ui_displayBusy();
+void ui_displayConfirm_init();
+void ui_displayScrollingText_init();
+
+void uiCallback_confirm(ui_callback_t* cb);
+void uiCallback_reject(ui_callback_t* cb);
 
 // responds to the host and resets
 // processing
 void respond_with_user_reject();
 
+#if defined(TARGET_NANOS)
+unsigned int ui_confirm_button(
+    unsigned int button_mask,
+    unsigned int button_mask_counter MARK_UNUSED
+);
+unsigned int ui_scrollingText_button(
+    unsigned int button_mask,
+    unsigned int button_mask_counter MARK_UNUSED
+);
+#endif
+
+extern displayState_t displayState;
+
+// WARNING(ppershing): Following two references MUST be declared `static`
+// otherwise the Ledger will crash. I am really not sure why is this
+// but it might be related to position-independent-code compilation.
+static scrollingState_t* scrollingState = &(displayState.scrolling);
+static confirmState_t* confirmState = &(displayState.confirm);
+
+enum {
+    INIT_MAGIC_SCROLLER = 2345,
+    INIT_MAGIC_CONFIRM = 5432,
+};
+
+static void ui_crash_handler() {
+#ifdef RESET_ON_CRASH
+    io_seproxyhal_se_reset();
+#else
+#endif
+}
+
+#define TRY_CATCH_UI(ui_call) \
```

Comment:
Nice, I like it.

Reply:
Hmm, actually, why is this only for nanos?

Reply:
it should be for both, the `if defined(TARGET_NANOS)` is ended at line 92 or am I missing something?